src/execution_plan/execution_plan_build/build_pattern_comprehension_ops.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2018-2020 Redis Labs Ltd. and Contributors
+ *
+ * This file is available under the Redis Labs Source Available License Agreement
+ */
+
+#include "execution_plan_construct.h"
+#include "RG.h"
+#include "../ops/ops.h"
+#include "../../query_ctx.h"
+#include "../execution_plan.h"
+#include "../../util/rax_extensions.h"
+#include "../execution_plan_build/execution_plan_modify.h"
+#include "../../arithmetic/arithmetic_expression_construct.h"
+
+// build pattern comprehension plan operations for example:
+// RETURN [p = (n)-->() | p] AS ps
+// Results
+//     Project
+//         Optional
+//             Aggregate
+//                 Conditional Traverse | (n)-[anon_0]->(anon_1)
+//                     All Node Scan | (n)
+//
+// MATCH (n) RETURN [p = (n)-->() | p] AS ps
+// Results
+//     Project
+//         Apply
+//             All Node Scan | (n)
+//             Optional
+//                 Aggregate
+//                     Conditional Traverse | (n)-[anon_1]->(anon_2)
+//                         Argument
+void buildPatternComprehensionOps(
+	ExecutionPlan *plan,
+	OpBase *root,
+	const cypher_astnode_t *ast
+) {
+	const cypher_astnode_t **pcs =
+		AST_GetTypedNodes(ast, CYPHER_AST_PATTERN_COMPREHENSION);
+	uint count = array_len(pcs);
+
+	const char **arguments = NULL;
+	if(root->childCount > 0) {
+		// get the bound variable to use when building the traversal ops
+		rax *bound_vars = raxNew();
+		ExecutionPlan_BoundVariables(root->children[0], bound_vars);
+		arguments = (const char **)raxValues(bound_vars);
+		raxFree(bound_vars);
+	}
+
+	for (uint i = 0; i < count; i++) {
+		const cypher_astnode_t *path =
+			cypher_ast_pattern_comprehension_get_pattern(pcs[i]);
+
+		QueryCtx_SetAST(plan->ast_segment);
```

Comment:
at the beginning of this function, keep a backup of the stored AST,
at the very end, restore this backup.